pkg/apis/discovery/v1beta1/conversion.go

```diff
@@ -0,0 +1,239 @@
+				continue
+			}
+
+			if _, ok := keyMap[k]; !ok {
+				keyMap[k] = 0
+				keysSize += len(k)
+			}
+
+			if _, ok := valueCounts[v]; !ok {
+				valueCounts[v] = 0
+				valuesSize += len(v)
+			}
+			valueCounts[v]++
+		}
+	}
+
+	approximateSize := approximateAnnotationSize(keyMap, valueCounts, len(in.Endpoints), keysSize, valuesSize)
+
+	maxSize := 256*(1<<10) - sizeOfAnnotations(out.Annotations)
+
+	removedStrings := 0
+	for approximateSize > maxSize {
+		keyToRemove := ""
+		for k := range keyMap {
+			if k != "kubernetes.io/hostname" && k != "topology.kubernetes.io/region" {
+				keyToRemove = k
+				break
+			}
+		}
+		// Either no more keys exist or remaining keys are standard keys
+		// No other way to preserve the data, so throw an error
+		if keyToRemove == "" {
+			return fmt.Errorf("unable to convert topology fields on v1beta1.EndpointSlice")
+		}
+		delete(keyMap, keyToRemove)
+		keysSize -= len(keyToRemove)
+		removedStrings++
+
+		for _, endpoint := range in.Endpoints {
+			val, ok := endpoint.Topology[keyToRemove]
+			if !ok {
+				continue
+			}
+			valueCounts[val]--
+			if valueCounts[val] == 0 {
+				delete(valueCounts, val)
+				valuesSize -= len(val)
+				removedStrings++
+			}
+		}
+
+		approximateSize = approximateAnnotationSize(keyMap, valueCounts, len(in.Endpoints), keysSize, valuesSize)
+	}
+
+	var strings []string
+	stringIndex := make(map[string]int)
+	var topologies []map[int]int
+	for _, endpoint := range in.Endpoints {
+		topology := make(map[int]int)
+		for k, v := range endpoint.Topology {
+			var kIndex, vIndex int
+			var ok bool
+			if kIndex, ok = stringIndex[k]; !ok {
+				if _, ok := keyMap[k]; !ok {
+					continue
+				}
+				kIndex = len(strings)
+				strings = append(strings, k)
+				stringIndex[k] = kIndex
+			}
+
+			if vIndex, ok = stringIndex[v]; !ok {
+				vIndex = len(strings)
+				strings = append(strings, v)
+			}
+			topology[kIndex] = vIndex
+		}
+		topologies = append(topologies, topology)
+	}
+
+	epsTopology := EndpointSliceTopology{Strings: strings, Topology: topologies}
+	topologyBytes, err := json.Marshal(epsTopology)
+	if err != nil {
+		return fmt.Errorf("errored marshaling endpoint slice topology fields: %q", err)
+	}
+
+	// If no strings, then topology is empty
+	if len(strings) == 0 {
+		return nil
+	}
+
+	if out.Annotations == nil {
+		out.Annotations = make(map[string]string)
+	}
+	out.Annotations[v1beta1Topology] = string(topologyBytes)
+
+	return nil
+}
+
+func Convert_discovery_EndpointSlice_To_v1beta1_EndpointSlice(in *discovery.EndpointSlice, out *v1beta1.EndpointSlice, s conversion.Scope) error {
+	// zone field conversion is taken care by endpoint conversion
+	if err := autoConvert_discovery_EndpointSlice_To_v1beta1_EndpointSlice(in, out, s); err != nil {
+		return err
+	}
+
+	annotation, ok := in.Annotations[v1beta1Topology]
+	if !ok {
+		return nil
+	}
+
+	var epsTop EndpointSliceTopology
+	if err := json.Unmarshal([]byte(annotation), &epsTop); err != nil {
```

Comment:
since this is a pretty high volume / high write rate API, are we concerned about the impact of this?

Reply:
use json-iterator?  Or keep it a field, as you suggested elsewhere